traversal/procedure/GraphProcedure.java

```diff
@@ -109,49 +110,41 @@ public int vertexCount() {
         return vertices.length;
     }
 
-    private void assertWithinFilterBounds(Set<Identifier.Variable.Retrievable> filter) {
-        assert iterate(vertices).anyMatch(vertex ->
-                vertex.id().isRetrievable() && filter.contains(vertex.id().asVariable().asRetrievable())
-        );
-    }
-
     @Override
     public FunctionalProducer<VertexMap> producer(GraphManager graphMgr, Traversal.Parameters params,
-                                                  Set<Identifier.Variable.Retrievable> filter, int parallelisation) {
+                                                  Modifiers modifiers, int parallelisation) {
         if (LOG.isTraceEnabled()) {
             LOG.trace(params.toString());
             LOG.trace(this.toString());
         }
-        assertWithinFilterBounds(filter);
-        if (initialVertex().id().isRetrievable() && filter.contains(initialVertex().id().asVariable().asRetrievable())) {
-            return async(initialVertex().iterator(graphMgr, params).map(v ->
-                    new GraphIterator(graphMgr, v, this, params, filter).distinct()
+        if (initialVertex().id().isRetrievable() && modifiers.filter().variables().contains(initialVertex().id().asVariable().asRetrievable())) {
+            return async(initialVertex().iterator(graphMgr, params, modifiers.sorting()).map(v ->
+                    new GraphIterator(graphMgr, v, this, params, modifiers).distinct()
```

Comment:
traversal execution now receives the modifiers. Openinga vertex iterator requires the sorting modifiers to decide whether to retrieve in ascending or descending order